src/serializer/ResidualHeapVisitor.js

```diff
@@ -498,8 +499,13 @@ export class ResidualHeapVisitor {
     });
   }
 
-  visitRoots(): void {
-    this.visitGenerator(this.realmGenerator);
+  // Use scopeOverride for setting the global generator to a function's scope
+  // instead of the realm generator.
+  visitRoots(scopeOverride?: Scope): void {
```

Comment:
Let's talk about `visitRoots`.

I think the way it should work is that it only gets called once, and gets some information about additional functions passed in. Within `visitRoots`, it would enumerate over the additional functions, apply (and later undo) effects, and visit what's relevant. Note that the fixpoint computation below should happen once, and over everything that's relevant.